File: Ruby/Endpoint Examples/JSON Payload/markdown.rb
#!
# What this sample does:
# - Converts a PDF to Markdown using pdfRest.
# - Uses a JSON payload in two steps: first uploads, then calls /markdown with the uploaded id.
#
# Setup (.env):
# - Copy .env.example to .env
# - Set PDFREST_API_KEY=your_api_key_here
# - Optional: set PDFREST_URL to override the API region. For EU/GDPR compliance and proximity, use:
#     PDFREST_URL=https://eu-api.pdfrest.com
#
# Usage:
#   ruby "Endpoint Examples/JSON Payload/markdown.rb" /path/to/input.pdf
#
# Output:
# - Prints the API JSON response to stdout. Non-2xx responses abort with a concise message.
# - Tip: pipe output to a file: ruby ... > response.json

require "json"
require "faraday"
require "faraday/retry"
require "dotenv"

Dotenv.load

API_KEY = ENV["PDFREST_API_KEY"]
abort("Missing PDFREST_API_KEY in .env") if API_KEY.nil? || API_KEY.strip.empty?

# Allow overriding the API base URL via .env (default to US endpoint)
API_BASE = (ENV["PDFREST_URL"] || ENV["PDFREST_API"] || "https://api.pdfrest.com").sub(%r{/+$}, "")

pdf_path = ARGV[0]
abort("Usage: ruby markdown.rb /path/to/file.pdf") unless pdf_path && File.file?(pdf_path)

filename = File.basename(pdf_path)
file_bytes = File.binread(pdf_path)

begin
  # 1) Upload the file
  upload_conn = Faraday.new(url: API_BASE) do |f|
    f.request :retry, max: 2, interval: 0.2
    f.adapter Faraday.default_adapter
  end

  upload_resp = upload_conn.post("/upload") do |req|
    req.headers["api-key"] = API_KEY
    req.headers["content-filename"] = filename
    req.headers["Content-Type"] = "application/octet-stream"
    req.body = file_bytes
  end

  STDERR.puts upload_resp.body
  unless upload_resp.success?
    abort("Upload failed with status #{upload_resp.status}")
  end

  upload_json = JSON.parse(upload_resp.body)
  uploaded_id = upload_json.fetch("files").fetch(0).fetch("id")
  STDERR.puts "Successfully uploaded with an id of: #{uploaded_id}"

  # 2) Convert to Markdown
  markdown_conn = Faraday.new(url: API_BASE) do |f|
    f.request :retry, max: 2, interval: 0.2
    f.adapter Faraday.default_adapter
  end

  body = { id: uploaded_id, page_break_comments: "on" }.to_json

  markdown_resp = markdown_conn.post("/markdown") do |req|
    req.headers["api-key"] = API_KEY
    req.headers["Content-Type"] = "application/json"
    req.body = body
  end

  puts markdown_resp.body
  unless markdown_resp.success?
    abort("Markdown conversion failed with status #{markdown_resp.status}")
  end

rescue KeyError => e
  abort("Unexpected response format: #{e.message}")
rescue => e
  abort("Error: #{e.class}: #{e.message}")
end

File: Perl/Endpoint Examples/JSON Payload/markdown.pl
#!/usr/bin/env perl
use strict;
use warnings;
use utf8;
use FindBin qw($Bin);
use File::Basename qw(basename);
use JSON::PP qw(encode_json decode_json);
use LWP::UserAgent;
use HTTP::Request;
use HTTP::Request::Common qw(POST);
use Encode qw(encode);
use Dotenv;

#!
# What this sample does:
# - Converts a PDF to Markdown using pdfRest.
# - Uses a JSON payload in two steps: upload to /upload, then call /markdown with the returned id.
#
# Setup (.env):
# - Copy .env.example to .env (Perl folder root)
# - Set PDFREST_API_KEY=your_api_key_here
# - Optional: set PDFREST_URL to override the API region. For EU/GDPR compliance and proximity, use:
#     PDFREST_URL=https://eu-api.pdfrest.com
# For more information visit https://pdfrest.com/pricing#how-do-eu-gdpr-api-calls-work
#
# Usage:
#   perl "Endpoint Examples/JSON Payload/markdown.pl" /path/to/input.pdf
#
# Output:
# - Prints the API JSON response to stdout. Non-2xx responses exit with a concise message.
# - Tip: pipe output to a file: perl ... > response.json

binmode STDOUT, ':raw';
binmode STDERR, ':encoding(UTF-8)';

# Load .env from the Perl folder root (two levels up from this script)
my $env_path = "$Bin/../../.env";
-e $env_path and Dotenv->load($env_path);

my $api_key = $ENV{PDFREST_API_KEY} // '';
if (!$api_key || $api_key =~ /^\s*$/) {
    print STDERR "Missing PDFREST_API_KEY in .env or environment\n";
    exit 1;
}

my $api_base = $ENV{PDFREST_URL} // $ENV{PDFREST_API} // 'https://api.pdfrest.com';
$api_base =~ s{/+$}{}; # trim trailing slashes

my $pdf_path = shift @ARGV;
if (!$pdf_path || !-f $pdf_path) {
    print STDERR "Usage: perl markdown.pl /path/to/file.pdf\n";
    exit 1;
}

my $filename = basename($pdf_path);
open my $fh, '<:raw', $pdf_path or do { print STDERR "Unable to read $pdf_path: $!\n"; exit 1; };
my $file_bytes; { local $/; $file_bytes = <$fh>; }
close $fh;

my $ua = LWP::UserAgent->new( timeout => 60 );

eval {
    # Step 1: Upload the file to receive a reusable id
    my $upload_req = HTTP::Request->new('POST', "$api_base/upload");
    $upload_req->header('api-key' => $api_key);
    $upload_req->header('content-filename' => $filename);
    $upload_req->header('Content-Type' => 'application/octet-stream');
    $upload_req->content($file_bytes);

    my $upload_resp = $ua->request($upload_req);
    print STDERR $upload_resp->decoded_content // '';
    if (!$upload_resp->is_success) {
        print STDERR "\nUpload failed with status " . $upload_resp->code . "\n";
        exit 1;
    }

    my $upload_json = decode_json($upload_resp->decoded_content // '{}');
    my $uploaded_id = $upload_json->{files} && ref $upload_json->{files} eq 'ARRAY'
        ? $upload_json->{files}[0]{id}
        : undef;
    if (!$uploaded_id) {
        print STDERR "Unexpected response format: missing files[0].id\n";
        exit 1;
    }
    print STDERR "Successfully uploaded with an id of: $uploaded_id\n";

    # Step 2: Request Markdown output using the uploaded id
    my $markdown_body = encode_json({ id => $uploaded_id, page_break_comments => 'on' });
    my $markdown_req = HTTP::Request->new('POST', "$api_base/markdown");
    $markdown_req->header('api-key' => $api_key);
    $markdown_req->header('Content-Type' => 'application/json');
    $markdown_req->content($markdown_body);

    my $markdown_resp = $ua->request($markdown_req);
    print STDOUT $markdown_resp->decoded_content // '';
    if (!$markdown_resp->is_success) {
        print STDERR "\nMarkdown conversion failed with status " . $markdown_resp->code . "\n";
        exit 1;
    }
    1;
} or do {
    my $err = $@ || 'Unknown error';
    $err =~ s/\s+$//;
    print STDERR "Error: $err\n";
    exit 1;
};

__END__
